### PR TITLE
[SPARK-6193] [EC2] Push group filter up to EC2

### PR DESCRIPTION
**Pull request by nchammas**
When looking for a cluster, spark-ec2 currently pulls down [info for all instances](https://github.com/apache/spark/blob/eb48fd6e9d55fb034c00e61374bb9c2a86a82fb8/ec2/spark_ec2.py#L620) and filters locally. When working on an AWS account with hundreds of active instances, this step alone can take over 10 seconds.

This PR improves how spark-ec2 searches for clusters by pushing the filter up to EC2.

Basically, the problem (and solution) look like this:

``` python
>>> timeit.timeit('blah = conn.get_all_reservations()', setup='from __main__ import conn', number=10)
116.96390509605408
>>> timeit.timeit('blah = conn.get_all_reservations(filters={"instance.group-name": ["my-cluster-master"]})', setup='from __main__ import conn', number=10)
4.629754066467285
```

Translated to a user-visible action, this looks like (against an AWS account with ~200 active instances):

``` shell
# master
$ python -m timeit -n 3 --setup 'import subprocess' 'subprocess.call("./spark-ec2 get-master my-cluster --region us-west-2", shell=True)'
...
3 loops, best of 3: 9.83 sec per loop

# this PR
$ python -m timeit -n 3 --setup 'import subprocess' 'subprocess.call("./spark-ec2 get-master my-cluster --region us-west-2", shell=True)'
...
3 loops, best of 3: 1.47 sec per loop
```

This PR also refactors `get_existing_cluster()` to make it, I hope, simpler.

Finally, this PR fixes some minor grammar issues related to printing status to the user. :tophat: :clap:

### CODE REVIEW

**Comment by nchammas** [comment]
cc @shivaram

File: ec2/spark_ec2.py [hunk 1]
```diff
@@ -612,40 +609,47 @@ def launch_cluster(conn, opts, cluster_name):
...
+        """
+        Get all non-terminated instances that belong to any of the provided security groups.
+
+        EC2 reservation filters and instance states are documented here:
+            http://docs.aws.amazon.com/cli/latest/reference/ec2/describe-instances.html#options
+        """
+        reservations = conn.get_all_reservations(
+            filters={"instance.group-name": group_names})
+        instances = itertools.chain.from_iterable(r.instances for r in reservations)
+        return [i for i in instances if i.state != "terminated"]
```
Comment (shivaram) [inline ec2/spark_ec2.py]: Minor comment: this seems a little different from the old check we had. We were originally checking if it was one of `'pending', 'running', 'stopping', 'stopped'` while right now we check if its not `terminated`. There are other states as shown in http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-lifecycle.html but I don't think it should matter.

**Comment by shivaram** [comment]
This is a nice change @nchammas -- Code looks good to me, but I'd just like to try it out once on my machine.

**Comment by shivaram** [comment]
Just tried this out and it works fine. LGTM pending the minor inline comment

File: ec2/spark_ec2.py (hunk 1 above)
Comment (nchammas) [inline ec2/spark_ec2.py]: The [doc referenced in the comment](http://docs.aws.amazon.com/cli/latest/reference/ec2/describe-instances.html#options) notes these instance states:

```
0 : pending
16 : running
32 : shutting-down
48 : terminated
64 : stopping
80 : stopped
```

So it looks like the only state we're missing is `shutting-down`, which is the intermediate state right before `terminated`. I can add that in to be consistent with the previous behavior.

Alternately, we can leave it as-is and re-terminate instances even if they are `shutting-down`. You know, zombies and stuff. :japanese_ogre: :gun:

Comment (shivaram) [inline ec2/spark_ec2.py]: Yeah it doesn't look like `shutting-down` is a major issue, but it might just be safer to keep existing behavior

Comment (nchammas) [inline ec2/spark_ec2.py]: OK, fixified.

**Comment by nchammas** [comment]
Inline comment addressed.

**Comment by srowen** [comment]
Sounds like a good improvement, changes look OK to my mildly informed eyes, and you have both reviewed and tested the change. LGTM.

**Pull request closed without merging**